### PR TITLE
Change edit an existing offer to use new flow

### PR DESCRIPTION
**Pull request by despo**
## Context

Enables the provider to edit the details of an existing offer made to a candidate.

### Clarification required: 
- [x] When is the offer tab displayed? What states can the application be in?
- [x] When should Defer offer link be displayed? In what application states?
- [x] How/When can conditions be marked as met? Map completed_offer_summary component to relevant actions.
- [ ] Clarify rules around text displayed on edit offer summary page (link to [Slack discussion](https://ukgovernmentdfe.slack.com/archives/CPH8J9G65/p1616516153113100?thread_ts=1616496274.102500&cid=CPH8J9G65)) **Out of scope**
- [x] An offer can only be edited in the offered state.
- [x] Enable all provider users to access offer tab, only ones with make decision can  

## Link to Trello card

https://trello.com/c/8Yc9CIab/3460-change-edit-an-existing-offer-to-use-new-flow

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Comment by adamsilver** [comment]
> When is the offer tab displayed? What states can the application be in?

The offer tab should be displayed when there's an offer associated with the application. Possible states include: offered, accepted, deferred, conditions met, conditions not met.

> When should Defer offer link be displayed? In what application states?

Only offers in the accepted or conditions met state can be deferred to the next cycle. See design history:
https://bat-design-history.netlify.app/manage-teacher-training-applications/deferring-offers-iteration/

> How/When can conditions be marked as met?

Conditions can only be marked as met or not met in the ‘accepted’ state.

>  Clarify rules around text displayed on edit offer summary page (link to Slack discussion)

We're deciding on the exact wording on the ‘check answers’ pages of the make and change offer flows today hopefully. Will let you know.

File: app/controllers/provider_interface/offer/study_modes_controller.rb [hunk 1]
```diff
@@ -25,6 +25,30 @@ def create
...
+        @wizard = OfferWizard.new(offer_store, study_mode_params.to_h)
+        @wizard.course_option_id = nil if @wizard.course_option_id && @wizard.course_option.study_mode != @wizard.study_mode
+
+        if @wizard.valid_for_current_step?
+          @wizard.save_state!
+
+          redirect_to [:edit, :provider_interface, @application_choice, :offer, @wizard.next_step]
+        else
+          @course = Course.find(@wizard.course_id)
+          @study_modes = avalable_study_modes(@course)
```
Comment (cwrw) [inline app/controllers/provider_interface/offer/study_modes_controller.rb:46]: ✂️  avalable_study_modes| available_study_modes

File: app/views/provider_interface/offers/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  <div class="govuk-grid-column-two-thirds">
+    <div class="govuk-heading-l govuk-!-margin-bottom govuk-!-font-size-36">Offer</div>
+      <% if provider_user_can_make_decisions %>
+        <p class="govuk-body govuk-!-margin-bottom-7">
+          <% if %i[pending_conditions recruited].include?(@application_choice.status.to_sym) %>
+            <%= govuk_link_to 'Defer offer', provider_interface_application_choice_new_defer_offer_path(@application_choice) %>
+          <% end %>
+
+          <% if %i[offer recruited].include?(@application_choice.status.to_sym) %>
+            <%= govuk_link_to 'Withdraw offer', provider_interface_application_choice_withdraw_offer_path(@application_choice) %>
```
Comment (cwrw) [inline app/views/provider_interface/offers/show.html.erb:12]: we might need a bit of spacing when both links are displayed when an application is recruited

**Review by cwrw — COMMENTED** [review]
Looks great! I have a few minor comments but otherwise 👍 
I tried testing all the transitions locally and they seem ok, only one Im not sure about is showing the withdraw link.

Another edge case I noticed, is when attempting to make an offer before the feature flag was set, and then switching the flag on I couldn't make an offer, even though it goes through the flow correctly, might be worth attempting to replicate. Let me know if you want to jump on a call for that

File: app/views/provider_interface/offers/show.html.erb [hunk 3]
```diff
@@ -3,11 +3,13 @@
   <div class="govuk-grid-column-two-thirds">
     <div class="govuk-heading-l govuk-!-margin-bottom govuk-!-font-size-36">Offer</div>
     <p class="govuk-body govuk-!-margin-bottom-7">
-      <% if %i[pending_conditions recruited].include?(@application_choice.status.to_sym)  %>
+      <% if %i[pending_conditions recruited].include?(@application_choice.status.to_sym) %>
         <%= govuk_link_to 'Defer offer', provider_interface_application_choice_new_defer_offer_path(@application_choice) %>
       <% end %>
 
-      <%= govuk_link_to 'Withdraw offer', provider_interface_application_choice_withdraw_offer_path(@application_choice) %>
+      <% if %i[offer recruited].include?(@application_choice.status.to_sym) %>
```
[the remark this answers is not in the document]
Comment (despo) [reply app/views/provider_interface/offers/show.html.erb]: Actually, we figured out yesterday when going through the process that only `offer` supports withdrawing an offer, which was a bit confusing as the existing code displays the link on both offer and recruited.

https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/services/application_state_change.rb#L57

File: app/views/provider_interface/offers/show.html.erb (hunk 2 above)
Comment (richardpattinson) [reply app/views/provider_interface/offers/show.html.erb:12]: After amending the conditions for displaying the link, defer and withdraw offer should now never appear at the same time so spacing isn't required.

**Review by cwrw — APPROVED** [review]
Looks good 👍  thanks for addressing comments. Last thing is the typo, after that is fixed happy for this to go in

File: app/controllers/provider_interface/offer/study_modes_controller.rb (hunk 1 above)
Comment (despo) [reply app/controllers/provider_interface/offer/study_modes_controller.rb]: 🩹

**Pull request merged**